### ISSUE TITLE
Global substitutions: substitute variables without adding duplicates

### ISSUE THREAD

**Issue by Emoun** [Emoun/duplicate#23]
### Short Description:

Make it possible to add substitution variables without adding more duplication of the code.

### Motivation:

A useful feature of this crate is the ability to reduce repetition without adding duplications:

```rust
#[duplicate(
  typ; [i32]
)]
fn some_func(in: &typ) -> typ{
  ..
}
```
Here we see how by only defining 1 duplication, we can reuse `typ` in the function.
This can be very useful where some piece of code is repeated many times in the same item.
Duplication is not needed, but we want to be able to insert some piece of code in many places.

While the above is currently possible, it cannot be used if duplication is also wanted.
In that case we would either need multiple calls to `duplicate` or  to repeat the body of the substitution variable for each duplication.

This proposal introduces first class support for _global substitutions_, which are substitutions that don't require duplication.
### Design

An insertion variable is a substitution variable that is only used for global substitution .
We declare an insertion variable using first an identifier, possibly followed by a parameter list, followed by a substitution and ended with semicolon:

```rust
#[duplicate(
  reference(type) [ & type];
)]
fn some_func(in: reference([i32])) -> reference([i64]){
  ..
}
```
This syntax is identical to how a substitution is declared in the verbose syntax and should therefore be natural to use and understand.
To increase the distinction between global substitution and duplication, we could enforce that all insertions must come at the beginning followed by the duplication declarations. If deemed unnecessary, this could be relaxed in the future without needing a major version bump.

### Misc:

* Should insertion variables be substituted in the bodies of succeeding insertion variables and the bodies of duplication substitution variables?

While this seems like a useful feature, allowing the use of insertion variables directly in the duplication declarations, it could also lead to confusion if the user doesn't intend to do it. Additionally, it also disallows the use of insertion variable identifiers in the duplication substitutions where insertions aren't wanted.
If we don't have this, reusing insertion variables would still be possible by putting all the needed insertions in a preceding call. This way the user can choose themselves whether they want to reuse the identifiers or the insertions.
Conclusion: No.

* Does this feature overlap with the use cases for nested invocation to a point where we could omit one of them?

While my original use case for nested invocations could be make use of insertions instead, nested invocations do have a power that cannot be replicated. Nested invocations allow exponential increases in duplication numbers by having one nested substitution group expand to multiple substitution groups in the parent.

**Comment by Emoun** [comment]
The proposed syntax results in ambiguity:

````rust
#[duplicate(
  reference [Ident];
)]
```

Is this an insertion variable or a parameterized substitution variable (in short syntax) missing any substitution group?

We must solve this before we can continue.

**Comment by Emoun** [comment]
The syntax ambiguity would be solved if #25 is implemented.

**Comment by Emoun** [comment]
Implemented to be released in next version.
